xc/device.ml

```diff
@@ -2059,30 +1885,334 @@ module Dm = struct
+          let parse_qmp_message = function
+            | Success (None, Vnc vnc) -> (try Some vnc.service with _ -> None)
+            | _ -> debug "Get unexpected result after sending Qmp message: %s" (string_of_message qmp_cmd); None
+          in
+          let qmp_cmd_result = qmp_write_and_read domid qmp_cmd in
+          match qmp_cmd_result with
+          | Some qmp_message -> parse_qmp_message qmp_message
+          | None -> debug "Fail to get result after sending Qmp message: %s" (string_of_message qmp_cmd); None
+        )
+
+      let is_hvm_linux domid =
+        let _XEN_IOPORT_LINUX_PRODNUM = 3 in (* from Linux include/xen/platform_pci.h *)
+        let error x = error "%s" x; Result.Error x in
+        try
+          let open Qmp in
+          match qmp_write_and_read domid (Command (None, Query_xen_platform_pv_driver_info)) with
+          | Some (Success (None, Xen_platform_pv_driver_info { product_num; build_num })) ->
+            Result.Ok ((product_num = _XEN_IOPORT_LINUX_PRODNUM) && (build_num <= 0xff))
+          | Some x ->
+            error (Printf.sprintf "Unexpected QMP response: %s" (Qmp.string_of_message x))
+          | None -> error "No QMP response for Query_xen_platform_pv_driver_info"
+        with e ->
+          error (Printf.sprintf "Exception attempting to obtain Linux product_id and build_number: %s" (Printexc.to_string e))
+
+      let maybe_write_pv_feature_flags ~xs domid =
+        match is_hvm_linux domid with
+        | Result.Ok true ->
+          let write_local_domain prefix x = xs.Xs.write (Printf.sprintf "/local/domain/%d/%s%s" domid prefix x) "1" in
+          List.iter (write_local_domain "control/feature-") ["suspend"; "shutdown"; "vcpu-hotplug"];
+          List.iter (write_local_domain "data/") ["updated"]
+        | _ -> ()
+
+      let suspend (task: Xenops_task.task_handle) ~xs ~qemu_domid domid =
+        Dm_Common.suspend task ~xs ~qemu_domid domid;
+        let file = sprintf qemu_save_path domid in
+        qmp_write domid (Qmp.Command(None, Qmp.Xen_save_devices_state file))
+
+      let init_daemon ~task ~path ~args ~name ~domid ~xs ~ready_path ?ready_val ~timeout ~cancel _ =
+        let pid = Dm_Common.init_daemon ~task ~path ~args ~name ~domid ~xs ~ready_path ?ready_val ~timeout ~cancel () in
+        Event.QMP_Event.add domid;
+        pid
+
+      let stop ~xs ~qemu_domid domid  =
+        Dm_Common.stop ~xs ~qemu_domid domid;
+        Event.QMP_Event.remove domid
+
+      let with_dirty_log domid ~f =
+        finally
+          (fun() ->
+             qmp_write domid (Qmp.Command(None, Qmp.Xen_set_global_dirty_log true));
+             f()
+          )
+          (fun() ->
+             qmp_write domid (Qmp.Command(None, Qmp.Xen_set_global_dirty_log false));
+          )
+
+    end (* Backend.Qemu_upstream_compat.Dm *)
+  end (* Backend.Qemu_upstream *)
+
+  (* Until stage 4, qemu_upstream behaves as qemu_upstream_compat *)
```

Comment:
Can you say something more for the uninitiated? ;-)

Reply:
done